src/printer.rs

```diff
@@ -159,79 +206,107 @@ impl Printer {
+                    self.buffer.print("\n")?;
                 }
             }
-            _ => {
-                if let Some(body) = get_body() {
-                    self.buffer.print(&body)?;
-                    self.buffer.print("\n\n")?;
-                }
-            }
-        };
+        }
         Ok(())
     }
 
-    pub async fn print_response_body(&mut self, mut response: Response) -> Result<()> {
-        match get_content_type(&response.headers()) {
-            Some(ContentType::Json) if self.stream => {
-                while let Some(bytes) = response.chunk().await? {
-                    self.print_json(&String::from_utf8_lossy(&bytes))?;
-                }
-            }
-            Some(ContentType::Xml) if self.stream => {
-                while let Some(bytes) = response.chunk().await? {
-                    self.print_xml(&String::from_utf8_lossy(&bytes))?;
-                }
-            }
-            Some(ContentType::Html) if self.stream => {
-                while let Some(bytes) = response.chunk().await? {
-                    self.print_html(&String::from_utf8_lossy(&bytes))?;
-                }
-            }
-            Some(ContentType::Json) => self.print_json(&response.text().await?)?,
-            Some(ContentType::Xml) => self.print_xml(&response.text().await?)?,
-            Some(ContentType::Html) => self.print_html(&response.text().await?)?,
-            _ => {
-                let mut is_first_chunk = true;
-                while let Some(bytes) = response.chunk().await? {
-                    if is_first_chunk
-                        && matches!(self.buffer, Buffer::Stdout | Buffer::Stderr)
-                        && bytes.contains(&b'\0')
-                    {
-                        self.buffer.print(BINARY_SUPPRESSOR)?;
-                        break;
-                    }
-                    is_first_chunk = false;
-
-                    self.buffer.write_all(&bytes)?;
-                }
+    pub fn print_response_body(&mut self, mut response: Response) -> anyhow::Result<()> {
+        if !self.buffer.is_terminal() {
+            // No trailing newlines, no binary detection, no decoding, direct streaming
+            self.print_body_stream(get_content_type(&response.headers()), &mut response)?;
+        } else if self.stream {
+            self.print_body_stream(
+                get_content_type(&response.headers()),
+                &mut decode_stream(&mut response),
+            )?;
+            self.buffer.print("\n")?;
+        } else {
+            let content_type = get_content_type(&response.headers());
+            // Note that .text() behaves like String::from_utf8_lossy()
+            let text = response.text()?;
+            if text.contains('\0') {
+                self.buffer.print(BINARY_SUPPRESSOR)?;
+                return Ok(());
             }
-        };
+            self.print_body_text(content_type, &text)?;
+            self.buffer.print("\n")?;
```

Comment:
Do we need this newline?

Reply:
Yes, otherwise if the output doesn't end in a newline it messes with the shell prompt. Sufficiently advanced shells have a [disgusting trick](https://www.vidarholen.net/contents/blog/?p=878) to fix it, but if you run `xh httpbin.org/json` with the current release in `bash` or `busybox ash` you get something like this at the end:
```
        "title": "Sample Slide Show"
    }
}~ $ 
```